File: src/codefresh/onprem.ts
import { fetchAndSaveData, prepareAndCleanup, RuntimeType, selectNamespace, writeCodefreshFiles } from '../deps.ts';

async function getAllAccounts(config: { headers: { Authorization: string }; baseUrl: string }) {
  const response = await fetch(`${config.baseUrl}/admin/accounts`, {
    method: 'GET',
    headers: config.headers,
  });
  const accounts = await response.json();
  await writeCodefreshFiles(accounts, 'onPrem-accounts');
}

async function getAllRuntimes(config: { headers: { Authorization: string }; baseUrl: string }) {
  const response = await fetch(`${config.baseUrl}/admin/runtime-environments`, {
    method: 'GET',
    headers: config.headers,
  });
  const onPremRuntimes = await response.json();
  await writeCodefreshFiles(onPremRuntimes, 'onPrem-runtimes');
}

async function getTotalUsers(config: { headers: { Authorization: string }; baseUrl: string }) {
  const response = await fetch(`${config.baseUrl}/admin/user?limit=1&page=1`, {
    method: 'GET',
    headers: config.headers,
  });
  const users = await response.json();
  await writeCodefreshFiles({ total: users.total }, 'onPrem-totalUsers');
}

async function getSystemFeatureFlags(config: { headers: { Authorization: string }; baseUrl: string }) {
  const response = await fetch(`${config.baseUrl}/admin/features`, {
    method: 'GET',
    headers: config.headers,
  });
  const onPremSystemFF = await response.json();
  await writeCodefreshFiles(onPremSystemFF, 'onPrem-systemFeatureFlags');
}

export async function onPrem(config: { headers: { Authorization: string }; baseUrl: string }) {
  if (config.baseUrl === 'https://g.codefresh.io/api') {
    console.error(
      `\nCannot gather On-Prem data for Codefresh SaaS. Please select either ${RuntimeType.pipelines} or ${RuntimeType.gitops}.`,
    );
    console.error(
      'If you need to gather data for Codefresh On-Prem, please update your ./cfconfig conext (or Envs) to point to an On-Prem instance.',
    );
    Deno.exit(40);
  }
  try {
    const namespace = await selectNamespace();
    console.log(`\nGathering data in "${namespace}" namespace for Codefresh On-Prem.`);
    await fetchAndSaveData(RuntimeType.onprem, namespace);
    await Promise.all([
      getAllAccounts(config),
      getAllRuntimes(config),
      getTotalUsers(config),
      getSystemFeatureFlags(config),
    ]);
    console.log('\nData Gathered Successfully.');
    await prepareAndCleanup();
  } catch (error) {
    console.error(`Error gathering On-Prem data:`, error);
  }
}

File: src/cf_support/commands/onprem.py
from ..models.onprem import OnPrem
from ..utils import files
import time


def execute(namespace):
    runtime_type = "onprem"
    dir_path = f"cf-support-{runtime_type}-{int(time.time())}"

    runtime = OnPrem()
    runtime.set_creds()

    if runtime.base_url == "https://g.codefresh.io/api":
        print(
            "Cannot gather On-Prem data for Codefresh SaaS. If you need to gather data for Codefresh On-Prem, please update your ./cfconfig context (or Envs) to point to an On-Prem instance."
        )
        return

    if not namespace:
        print(f"Which namespace is Codefresh On-Prem installed in?")
        runtime.select_namespace()
    else:
        runtime.namespace = namespace

    print(f"Gathering data in the {runtime.namespace} namespace")
    k8s_resources = runtime.get_k8s_resources()
    files.save_k8s_resources(k8s_resources, dir_path)

    if runtime.base_url != None:
        accounts = runtime.get_accounts()
        files.save_file(files.to_yaml(accounts), "onprem_accounts.yaml", dir_path)

        runtimes = runtime.get_runtimes()
        files.save_file(files.to_yaml(runtimes), "onprem_runtimes.yaml", dir_path)

        total_users = runtime.get_total_users()
        files.save_file(total_users, "onprem_total_users.txt", dir_path)

        features = runtime.get_feature_flags()
        files.save_file(files.to_yaml(features), "onprem_features.yaml", dir_path)

    print("Gathering data completed")
    files.compress_dir(dir_path)
